### ISSUE TITLE
case class 22 argument limit

### ISSUE THREAD

**Issue by polkovnikov-ph** [slick/slick#1462]
Codegenerator [still checks](https://github.com/slick/slick/blob/9f7784ae5837e483815d6b45b9de289c6d4f817d/slick-codegen/src/main/scala/slick/codegen/AbstractGenerator.scala#L90) if there're less than 22 fields in a table to generate a case class, while in Scala 2.11 this issue was already fixed. Could we drop this line or enable such check explicitly via an extra argument?

**Comment by cvogt** [comment]
@polkovnikov-ph unfortunately not just like that, because case classes >22 don't have apply and unapply methods even in > 2.11. We could generate them and I am hoping we'll find a capable Google Summer of Code student over the summer to implement it.

**Comment by Petesta** [comment]
@cvogt I thought the limit on case classes was lifted in >= 2.11? You're able to define case classes with more than 22 fields and still have apply available on it. So couldn't the limit then be removed as said by @polkovnikov-ph?

```
scala> case class Foo(a: Int, b: Int, c: Int, d: Int, e: Int, f: Int, g: Int, h: Int, i: Int, j: Int, k: Int, l: Int, m: Int, n: Int, o: Int, p: Int, q: Int, r: Int, s: Int, t: Int, u: Int, v: Int, w: Int, x: Int, y: Int, z: Int)

scala> Foo(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)
res1: Foo = Foo(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)
```

**Comment by cvogt** [comment]
@Petesta Oh, I see I had a misconception there. You are right they DO have apply and unapply methods in scala 2.11. BUT they can't be turned into functions, which you'd need to make slick with with them.

```
scala> Foo.apply _
<console>:13: error: implementation restricts functions to 22 parameters
       Foo.apply _
```

**Comment by mdedetrich** [comment]
Yeah, until when (and if) Dotty fixes this issue, having functions/tuples of any arity in Scala isn't really possible. You can replace the arity limit of tuples with a `HList` implementation, which is what you need to do currently if you want to bypass the 22 arity limit, but `HList` are a more restrictive version of tuples, which is what Shapeless uses by default

I think there is another task for better interopt for Shapeless with Slick, which would make this issue less painful (you would use Shapeless `Generic` instead of `.tupled`/`.unapply`)

**Comment by mb73** [comment]
Hello :) I forked slick to solve this issue for my use case and would like to discuss if my solution is suitiable for slick:master too. I outlined my approach here: https://github.com/slick/slick/issues/1885

In summary for tables with more than 22 columns, I suggest to keep the `HList` for plain SQL fetching (because there is no `unapply `for large case classes) and have the `case class` for the rest. This works fine for me. 

There is a limition though: To compare what was fetched by plain SQL (a `HList`) with instances of the `case class` you have to do some conversion, e.g. (from `CodeGenRoundTipTest`)
`          sql"""select * from #${quoteIdentifier("LARGE")} where #${quoteIdentifier("id")} = 0""".as[LargeRowHList].head.map(res => assertEquals(oData.productIterator.toList, res.toList))
`

What do you think? 

Note: I already created a PR, but noticed afterwards that `CodeGenRoundTipTest` was broken. [I have fixed that](https://github.com/mb73/slick/commit/34b68c5d805520de78fb04124acff3f6d71104bd) but still have to add move fix to the PR (no idea how to, though).

**Comment by hvesalai** [comment]
@mb73 so how would using a table with 23 columns look like in lifted? How do you implement the `<>` mapping for it, for example?

**Comment by mb73** [comment]
I just use a `toMap` macro. 

Small rows table:
`def * = (id, title, category).mapTo[PostsRow]`

Large rows table:
`def * = (id :: p1i1 :: p1i2 :: p1i3 :: p1i4 :: p1i5 :: p1i6 :: p2i1 :: p2i2 :: p2i3 :: p2i4 :: p2i5 :: p2i6 :: p3i1 :: p3i2 :: p3i3 :: p3i4 :: p3i5 :: p3i6 :: p4i1 :: p4i2 :: p4i3 :: p4i4 :: p4i5 :: p4i6 :: p5i1 :: p5i2 :: p5i3 :: p5i4 :: p5i5 :: p5i6 :: p6i1 :: p6i2 :: p6i3 :: p6i4 :: p6i5 :: p6i6 :: HNil).mapTo[LargeRow]`

**Comment by mb73** [comment]
Using a table with 23 columns is exactly the same as with 22 columns. At least I cannot find any difference. As for `<>` just use `mapTo` instead as described above. The rest stays the same.

**Comment by mb73** [comment]
I started my PR from scratch. My new solution is without the limitations mentioned above (has `def ?` and also returns `case class`es using plain SQL. Also now only uses mapTo when needed (maybe better compile times with less macro).

**Comment by mb73** [comment]
Generated option for large rows table (`slick.test.codegen.roundtrip.Tables.LargeRow`):

`def ? = (Rep.Some(id) :: Rep.Some(p1i1) :: Rep.Some(p1i2) :: Rep.Some(p1i3) :: Rep.Some(p1i4) :: Rep.Some(p1i5) :: Rep.Some(p1i6) :: Rep.Some(p2i1) :: Rep.Some(p2i2) :: Rep.Some(p2i3) :: Rep.Some(p2i4) :: Rep.Some(p2i5) :: Rep.Some(p2i6) :: Rep.Some(p3i1) :: Rep.Some(p3i2) :: Rep.Some(p3i3) :: Rep.Some(p3i4) :: Rep.Some(p3i5) :: Rep.Some(p3i6) :: Rep.Some(p4i1) :: Rep.Some(p4i2) :: Rep.Some(p4i3) :: Rep.Some(p4i4) :: Rep.Some(p4i5) :: Rep.Some(p4i6) :: Rep.Some(p5i1) :: Rep.Some(p5i2) :: Rep.Some(p5i3) :: Rep.Some(p5i4) :: Rep.Some(p5i5) :: Rep.Some(p5i6) :: Rep.Some(p6i1) :: Rep.Some(p6i2) :: Rep.Some(p6i3) :: Rep.Some(p6i4) :: Rep.Some(p6i5) :: Rep.Some(p6i6) :: HNil).shaped.<>(r => LargeRow(r(0).asInstanceOf[Option[Long]].get, r(1).asInstanceOf[Option[Int]].get, r(2).asInstanceOf[Option[Int]].get, r(3).asInstanceOf[Option[Int]].get, r(4).asInstanceOf[Option[Int]].get, r(5).asInstanceOf[Option[Int]].get, r(6).asInstanceOf[Option[Int]].get, r(7).asInstanceOf[Option[Int]].get, r(8).asInstanceOf[Option[Int]].get, r(9).asInstanceOf[Option[Int]].get, r(10).asInstanceOf[Option[Int]].get, r(11).asInstanceOf[Option[Int]].get, r(12).asInstanceOf[Option[Int]].get, r(13).asInstanceOf[Option[Int]].get, r(14).asInstanceOf[Option[Int]].get, r(15).asInstanceOf[Option[Int]].get, r(16).asInstanceOf[Option[Int]].get, r(17).asInstanceOf[Option[Int]].get, r(18).asInstanceOf[Option[Int]].get, r(19).asInstanceOf[Option[Int]].get, r(20).asInstanceOf[Option[Int]].get, r(21).asInstanceOf[Option[Int]].get, r(22).asInstanceOf[Option[Int]].get, r(23).asInstanceOf[Option[Int]].get, r(24).asInstanceOf[Option[Int]].get, r(25).asInstanceOf[Option[Int]].get, r(26).asInstanceOf[Option[Int]].get, r(27).asInstanceOf[Option[Int]].get, r(28).asInstanceOf[Option[Int]].get, r(29).asInstanceOf[Option[Int]].get, r(30).asInstanceOf[Option[Int]].get, r(31).asInstanceOf[Option[Int]].get, r(32).asInstanceOf[Option[Int]].get, r(33).asInstanceOf[Option[Int]].get, r(34).asInstanceOf[Option[Int]].get, r(35).asInstanceOf[Option[Int]].get, r(36).asInstanceOf[Option[Int]].get), (_:Any) =>  throw new Exception("Inserting into ? projection not supported."))`

Would be nice if someone could test it. I don't have any outer joins in my use case ...

**Comment by mb73** [comment]
Plain SQL also returns a case class instance. Again from `slick.test.codegen.roundtrip3.Tables.Large`:

```
  implicit def GetResultLargeRow(implicit e0: GR[Long], e1: GR[Int]): GR[LargeRow] = GR{
    prs => import prs._
    LargeRow(<<[Long], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int], <<[Int])
  }
```

And here the generated class itself:

```
  case class LargeRow(id: Long, p1i1: Int = 11, p1i2: Int = 12, p1i3: Int = 13, p1i4: Int = 14, p1i5: Int = 15, p1i6: Int = 16, p2i1: Int = 21, p2i2: Int = 22, p2i3: Int = 23, p2i4: Int = 24, p2i5: Int = 25, p2i6: Int = 26, p3i1: Int = 31, p3i2: Int = 32, p3i3: Int = 33, p3i4: Int = 34, p3i5: Int = 35, p3i6: Int = 36, p4i1: Int = 41, p4i2: Int = 42, p4i3: Int = 43, p4i4: Int = 44, p4i5: Int = 45, p4i6: Int = 46, p5i1: Int = 51, p5i2: Int = 52, p5i3: Int = 53, p5i4: Int = 54, p5i5: Int = 55, p5i6: Int = 56, p6i1: Int = 61, p6i2: Int = 62, p6i3: Int = 63, p6i4: Int = 64, p6i5: Int = 65, p6i6: Int = 66)
```

**Comment by hvesalai** [comment]
Solved by @mb73 in #1889

**Comment by hvesalai** [comment]
@polkovnikov-ph never too late
